applications/NurbsBrepApplication/custom_utilities/geometries/NodeSurfaceGeometry3D.h

```diff
@@ -0,0 +1,140 @@
+/*
+//  KRATOS _____________ 
+//        /  _/ ____/   |
+//        / // / __/ /| |
+//      _/ // /_/ / ___ |
+//     /___/\____/_/  |_| StructuralMechanicsApplication
+//
+//  Author: Thomas Oberbichler
+*/
+
+#pragma once
+
+#include <ANurbs/Core>
+
+
+namespace Kratos {
+
+class NodeSurfaceGeometry3D
+    : public ANurbs::SurfaceGeometryBase<double, ANurbs::Point3D>
+{
+protected:
+    using typename NodePointer = typename Node<3>::Pointer;
+
+public:
+    using SurfaceGeometryBaseType = SurfaceGeometryBase<double, ANurbs::Point3D>;
+    using typename SurfaceGeometryBaseType::KnotsType;
+    using typename SurfaceGeometryBaseType::ScalarType;
+    using typename SurfaceGeometryBaseType::VectorType;
+
+protected:
+    ANurbs::Grid<NodePointer> m_nodes;
+
+public:
+    NodeSurfaceGeometry3D(
+        const int& degreeU,
```

Comment:
If it is const you can remoce the &

Reply:
I mean, it is the same when using int, bool and double

Reply:
Maybe old fashioned PassByRef. Just saw it in the other parts of Kratos

Reply:
[Would you keep the const?](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f16-for-in-parameters-pass-cheaply-copied-types-by-value-and-others-by-reference-to-const)

Reply:
Yes, yes keep it const. The thing is that a reference is a pointer, then it takes the same space than a  double. So in the case of a Vector or a Matrix you are using just "a double", but when you are using a bool, int or double you are using more. The reference is necessary in case you want to modify the value, but in case it is const it is not. In any case the compiler knows this and optimizes this part of the code, but you cannot trust compilers....

Reply:
I recommend to use std::size_t instead of int in case the integer is just positive

Reply:
When I started the project I was using `size_t` everywhere. But now I am sure that **unsigned integers are the work of the devil!**

The problem is that a lot of Nurbs algorithms are expecting integers. [They become very buggy because of unnoticed overflows.](https://youtu.be/Puio5dly9N8?t=42m40s)

So I would like to add some lines to check e.g. degree > 0 (btw: that would also be necessary for `size_t`).

Reply:
I see